### PR TITLE
feat: React Chrono, BAT indicator, rudder trim indicator

### LINKED ISSUE #3237: [REQUEST] 【Clock】 When the time is stopped by pressing the CHR button, there is no colon in center
**Is your feature request related to a problem? Please describe.**

Clock enhancements

**Describe the feature you'd like to see implemented**

I have a request regarding the chrono counter (CHR) function of the clock.
When the time is stopped by pressing the CHR button, the ":" is not displayed in the centres.
The ":" is only displayed if the clock is running.

**References**

WS000007

VIDEO (1:00~)
*There is backlighting and it is difficult to see.

https://www.youtube.com/watch?v=xczl6YiLRl0&t=70s

**Additional context**

I have researched FCOM. But for some reason I could not find a description.
So the only reference I have is a video. Sorry.

### LINKED ISSUE #3238: [REQUEST] 【Clock】Function to stop or reset the Elapsed Time (ET)
**Is your feature request related to a problem? Please describe.**

Clock enhancements

**Describe the feature you'd like to see implemented**

I would like a function to stop or reset the elapsed time (ET).

**References**

VIDEO (2:15~)
*There is backlighting and it is difficult to see.
https://youtu.be/xczl6YiLRl0?t=138

FCOM
ET0
ET1
ET2

**Additional context**

I hope that one day it will be implemented.

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by lukecologne**
Fixes #3237
Fixes #3238
Replaces #3469

## Summary of Changes

- [x] Reimplement BAT indicator in react
- [x] Reimplement RTPI in react
- [x] Reimplement Chrono in react
   - [x] Stopwatch
   - [x] Clock/Date - **Note: switching from GPS time to internal time, and changing the internal time is not implemented yet.**
   - [x] Elapsed Time
      - [x] Code
      - [x] XML - **Note: the elapsed time knob should be springloaded in the 3rd position (reset). This is not yet implemented. If someone has an idea of how to do this, feel free to make a suggestion.**
* The visuals should be pretty much identical. 

## References

* Chrono: https://youtu.be/xczl6YiLRl0
* RTPI: https://www.youtube.com/watch?v=hRWg4g3UiHo

## Additional context

Discord username (if different from GitHub):

## Testing instructions

* Make sure the chrono and RTPI behave according to the references provided. (Except the issue described above)
* Make sure the BAT indicator behaves properly.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by lukecologne** [comment]
Should be ready for review. It still has one issue: the elapsed time knob should be springloaded in the 3rd position (reset). This is not yet implemented. If someone has an idea of how to do this, feel free to make a suggestion.

File: src/instruments/src/Clock/Components/Chrono.tsx
Comment (tracernz) [inline src/instruments/src/Clock/Components/Chrono.tsx:18]: Is this robust to sim time being changed e.g. in the weather pop-up? P3D docs say `ABSOLUTE TIME` is:  
`Time, as referenced from 12:00 AM January 1, 0000`  
I don't know if there is a monotonic clock simvar available.

**Comment by beheh** [comment]
#3469 has landed. Could you please make sure to incorporate any changes from it, if you haven't already? Thanks!

File: src/instruments/src/Clock/Components/Chrono.tsx
Comment (beheh) [reply src/instruments/src/Clock/Components/Chrono.tsx:18]: So according to some research by @theomessin, Coherent guarantees well-behaved and monotone behaviour when using the native `setTimeout` and `setInterval`. If `ABSOLUTE TIME` is not the time from outside of the machine we could either do what deltaTime does and use `Date.now()`, or `setTimeout`/`setInterval` respectively.

**Comment by lukecologne** [comment]
> 
> 
> #3469 has landed, I only saw this afterwards. Could you please make sure to incorporate any changes from it, if you haven't already? Thanks!
> 
> Edit: Also there's #3448 - is that incorporated already too?

I just reimplemented the Chrono after the real life refernce, and that includes the change from this PR. Not sure what #3448 is supposed to do, the screenshot isn't really informative. I will resolve the conflicts though

File: src/instruments/src/Clock/Components/Chrono.tsx
Comment (lukecologne) [reply src/instruments/src/Clock/Components/Chrono.tsx:18]: The good thing with using ABSOLUTE TIME is that it respects variable sim rate, which is pretty important IMO, since the real-time isn't really of interest when using the stopwatch/elapsed time.

**Comment by lukecologne** [comment]
Ok, conflicts are resolved.

**Comment by beheh** [comment]
#3448 is about some CSS changes - maybe @Snapmatics could chime in or clarify what changes that are? Otherwise #3448 might have to be completely redone after this lands.

**Comment by lukecologne** [comment]
The only change that is not either whitespace or formatting is this: https://github.com/flybywiresim/a32nx/pull/3448/files?diff=split&w=1#diff-6117fb7e0993a922fb071577fcfd106edc6b554a8de8f97cb7f7237bd2a48257L70. I added a reference for the RTPI to my PR comment, the testers should compare it to my PR and I can adjust it if something looks off.

**Comment by wpine215** [comment]
is this ready for QA testing?

**Comment by lukecologne** [comment]
From me, yes. Obviously someone with react knowledge needs to review first.

File: src/instruments/src/Clock/Components/Chrono.tsx
Comment (beheh) [reply src/instruments/src/Clock/Components/Chrono.tsx:19]: Alright, sounds good. I'm sure QA will test this.

**Review by beheh — APPROVED** [review]

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 17.3.2021
Version of the sim: 1.14.5.0
PR Tested: #3794
Tier of testing: 1
Changes to observe: BAT, RTPI and Chrono
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Looked at Bat displays
3.) Tested Chrono (Image1+2)
4.) Tested RTPT

Media:


Image1


Image2

Issues: Last two digits from timer and ET jump from colon and are cut off

Overall Rating: Needs to be improved

Conclusions of the testing: One issue, rest works fine. Chrono and RTPI displays are a little dim in daylight in my opinion.

Hope this helps :-)

**Comment by marcman86** [comment]
> 
> 
> Category: QA Tester
> Name: marcman86#4907
> Date of testing: 17.3.2021
> Version of the sim: 1.14.5.0
> PR Tested: #3794
> Tier of testing: 1
> Changes to observe: BAT, RTPI and Chrono
> Testing technique: Tryout
> 
> Steps made to test the Pull Request:
> 
> 1.) Spawned C&D at EDDP
> 2.) Looked at Bat displays
> 3.) Tested Chrono (Image1+2)
> 4.) Tested RTPT
> 
> Media:
> 
> 
> Image1
> 
> 
> Image2
> 
> Issues: Last two digits from timer and ET jump from colon and are cut off
> 
> Overall Rating: Needs to be improved
> 
> Conclusions of the testing: One issue, rest works fine. Chrono and RTPI displays are a little dim in daylight in my opinion.
> 
> Hope this helps :-)

Retested after fix: Looks good now

**Comment by DavidEGTE** [comment]
Category: QA Tester
Name: David#2177
Date of testing: 17/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3794 
Tier of testing: 1

Changes to observe: BAT, RTPI and Chrono

Testing technique: Try-out

Steps you've made to test the Pull Request:
1. Spawned C&D
2. Tested BAT (seems to be working)
3. Tested RTPI (photo below + issue)
4. Tested Chrono (photo below)

Flight notes (When done a full flight):
N/A

Media: 
Microsoft Flight Simulator Screenshot 2021 03 17 - 13 01 00 03
Microsoft Flight Simulator Screenshot 2021 03 17 - 13 11 49 60

Issues (If Any): Write any issues
Shouldn't the letter "r" be in uppercase?
Probably not relative on this PR, shouldn't reset button decrease to 0.0 on RTPI

Steps to achieve the issue:
N/A

Overall Rating: Good.
Conclusions of the testing: Small issue but LGTM

**Comment by MisterChocker** [comment]
sry mb, one more test missing.

**Comment by lukecologne** [comment]
> Issues (If Any): Write any issues
> Shouldn't the letter "r" be in uppercase?
> Probably not relative on this PR, shouldn't reset button decrease to 0.0 on RTPI

Thank you for the review. The problem with the lowercase R is in the font, nothing I can do there, I already use an uppercase R in the code. The rudder trim reset is unrelated, this PR only changed the display, nothing on the actual knob/button

**Comment by Matse08** [comment]
Category: QA Tester
Name: Matse#7454
Date of testing: 19.02.21
Version of the sim: 1.14.5.0
PR Tested: #3794
Tier of testing: 2
Changes to observe:
BAT
RTPI
Chrono


Testing technique: Try-out
Steps you've made to test the Pull Request:
1. Spawned cold and dark
2. Tested RTPI 
3. Tested Chrono 
4. Tested BAT


Flight notes (When done a full flight):N/A


Issues (If Any): No issue


Overall Rating: Good
Conclusions of the testing: This does work great for me

**Comment by lukecologne** [comment]
Don't merge, stupid me forgot the changelog

**Comment by MisterChocker** [comment]
you are 1 seconds to late -.-

**Pull request merged**
